### PR TITLE
Add nuclear zone data

### PR DESCRIPTION
**Pull request by whokion**
# Extend neutron inelastic data for the Bertini cascade nuclei model

This MR adds data for nuclear zone properties to NeutronInelasticData and associated updates:
 - Import nucleon loss energies in IsotopeView
 - Add options for the Bertini cascade model
 - Add nuclear zone data for the Bertini cascade model
 - Extend the NeutronInelastic test and adjust other tests due to updated in Isotope

### CODE REVIEW

**Comment by whokion** [comment]
@sethrj Is the fail from windows real?

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: @whokion This is the source of the windows error:
```
D:\a\celeritas\celeritas\src\celeritas\neutron\model\detail/NuclearZoneBuilder.hh(368): error C2220: the following warning is treated as an error
2024-06-12T13:55:56.9194698Z D:\a\celeritas\celeritas\src\celeritas\neutron\model\detail/NuclearZoneBuilder.hh(368): warning C4244: 'initializing': conversion from 'double' to 'const int', possible loss of data
```

Write it as `1000`. (In theory you could specify a real number that's out of range of `int` and then loses accuracy.)

Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Thanks!

File: src/celeritas/ext/GeantImporter.cc
Comment (sethrj) [inline src/celeritas/ext/GeantImporter.cc]: This is causing the test failure:
```
error: expected a JSON object but got the following stdout:
G4NucleiProperties::GetMassExccess: Wrong values for A = 0 and Z = 0
G4NucleiProperties::GetMassExccess: Wrong values for A = 0 and Z = 1
```

We probably should add 
```c++
    ScopedGeantLogger scoped_logger;
```
to  `GeantImporter::operator()`.

I guess you'll have to guard these calls with `A > 1` and `Z > 1`?

Comment (whokion) [reply src/celeritas/ext/GeantImporter.cc]: Okay, thanks!  This still looks a bit strange as it should return 0 and exit from [G4NucleiProperties::GetBindingEnergy](https://geant4.kek.jp/lxr/source/particles/management/src/G4NucleiProperties.cc#L238) for those cases and should not call `G4NucleiProperties::GetMassExccess`.

File: src/celeritas/ext/GeantImporter.cc
Comment (sethrj) [inline src/celeritas/ext/GeantImporter.cc]:
```suggestion
        // Binding energy difference for losing a nucleon
```

File: src/celeritas/io/ImportElement.hh
Comment (sethrj) [inline src/celeritas/io/ImportElement.hh]: All of the data in this struct would benefit from default initializers (missed that in earlier reviews)

File: src/celeritas/neutron/model/CascadeOptionsIO.json.cc
Comment (sethrj) [inline src/celeritas/neutron/model/CascadeOptionsIO.json.cc]: There's a little utility function that makes this easier (and slightly more efficient and slightly more flexible):

```suggestion
#define FDO_INPUT(NAME)  CELER_JSON_LOAD_OPTION(j, opts, NAME)
```
and include "corecel/io/JsonUtils.json.hh"

File: src/celeritas/neutron/model/NeutronInelasticModel.cc
Comment (sethrj) [inline src/celeritas/neutron/model/NeutronInelasticModel.cc]: Rename the instance to the function-like `build_zone` :)

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Doesn't need to be explicit (since there's more than one construction argument)

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Might be a bit clearer as a couple of chained `if` statements, what do you think?
```suggestion
    if (a < AtomicMassNumber{5})
        return 1;
    if (a < AtomicMassNumber{100})
        return 3;
    return 6;
```

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: This and the `integral` clear are unnecessary as they'll be destroyed at the end of this function

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: It would be a little less error prone to calculate the total after the big `if` statements, rather than accumulating in four different places:
```c++
real_type total_integral = std::accumulate(integral.begin(), integral.end(), real_type{0});
```

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:362]: This function is deep enough and complicated enough to merit a standalone class with testing, and maybe configurable parameters like the maximum trial depth. (Note also that a linear increase in the "max trials" results in a *quadratic* increase in the execution time due to the inner loop.)

If you make a function-like class `PotentialIntegrator` these could be construction arguments... and since this looks like it could be a generic recursive integrator, I think it would be even better as a standalone testable `NewtonCotesIntegrator` class in `corecel/math` templated on `class F` (which evaluates the potential). And I don't think the below is a very optimal implementation of it...

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: I think this is wrong: `false` is being promoted to `rmin = 0`, `rmin` is being passed as `delta_r`, etc.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: I think this would be better written with `SoftEqual`? (Use a function-local `constexpr SoftEqual<>{epsilon}`). The SoftEqual also has an absolute tolerance that might be useful (or might be harmful); what's the scale of the expected result?

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:321]: This (and the gaussian case below) are just initial guesses for the integral... this should be trivial to incorporate into the adaptive quadrature solver itself, right?

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: This would be easier to read with 
```c++
    else if (amass < 12)
    {
        constexpr Real3 sqrt_minus_log_alpha = {/* precompute these*/};
        /* * */
    }
    else if (amass < 100)
    {
        constexpr Real3 alpha = {0.7, 0.3, 0.01};
        /* * */
    }
```

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: I don't think you need a whole vector of `ur`: only `[i, i+1]` are used at a time...

**Review by sethrj — CHANGES_REQUESTED** [review]
Everything but the NuclearZoneBuilder looks essentially perfect! great design and implementation.

I think the NZB needs to split out a templated numerical integrator class that has some additional testing on it.

File: src/celeritas/ext/GeantImporter.cc
Comment (whokion) [reply src/celeritas/ext/GeantImporter.cc]: Fixed.

File: src/celeritas/io/ImportElement.hh
Comment (whokion) [reply src/celeritas/io/ImportElement.hh]: Added  default initializers.

File: src/celeritas/neutron/model/CascadeOptionsIO.json.cc
Comment (whokion) [reply src/celeritas/neutron/model/CascadeOptionsIO.json.cc]: Updated.

File: src/celeritas/neutron/model/NeutronInelasticModel.cc
Comment (whokion) [reply src/celeritas/neutron/model/NeutronInelasticModel.cc]: Rename build_zone to build_nuclear_zones.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Removed explicit.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Okay, changed as suggested.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Removed.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Good suggestion!  Updated accordingly.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:362]: Great suggestion - actually I thought about how to pass a functor like object to `integrate_potential`, but did not think harder^^.  If you do not mind, I would like to do this as a separate MR or you may help me here as you already have a good idea.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Nice catch!  This is definitely a bug (left over from the earlier implementation).  I also realize that this Gaussian potential function is not tested (i.e, any A in [5, 11]), so the bug was not caught in validation (with respect to Geant4) in the test.  I will add a test case.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Updated as suggested - at least, good to see that there is no changes test results.

File: src/celeritas/neutron/data/NeutronInelasticData.hh
Comment (amandalund) [inline src/celeritas/neutron/data/NeutronInelasticData.hh]:
```suggestion
```
Since we can get the number from `zones.size()`, right?

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:321]: You are absolutely right. This was a lazy man's implementation, but definitely can be integrated into `PotentialIntegrator` that you suggested early.

File: src/celeritas/neutron/data/NeutronInelasticData.hh
Comment (amandalund) [inline src/celeritas/neutron/data/NeutronInelasticData.hh:98]: I think the data could be simplified slightly by removing this struct and instead storing the `components` and `zones` directly in `NeutronInelasticData`.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Okay, I did not through about this pre-computation during refactoring Geant4 codes.  Good suggestion. Changed accordingly.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Nice catch! Removed this vector, but use a temporary variable (one double to keep the last value of y).

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (amandalund) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]:
```suggestion
```

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (amandalund) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]:
```suggestion
```

**Comment by whokion** [comment]
@sethrj Thanks a lot for your review comments and suggestions, which I hope that I addressed most of them except the request for PotentialIntegrator (which may be a separate MR).  Is the CI failure something that I should worry about or a just glitch?

**Comment by sethrj** [comment]
@whokion Just a glitch I think: rerunning now. I'll see if I can make a small PR with a generic numerical integrator and test...

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:171]: Instead of doing the "number of zones"  calculation in a separate function and expecting the size to match in the `if` statements below, I think it'd be better to resize the vector inside this function. Inside the `if` statements, resize the vector to 1, `y.size()`, `alpha.size()`, etc., and then loop over `integral.size()` rather than a separate `num_of_zones` values.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: `num_zones` is redundant with `zones.size()`, so maybe that data should be removed?

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]:
```suggestion
    int num_protons = target.atomic_number().get();
    int const num_nucleons[] = {num_protons, a.get() - num_protons};
    real_type const mass[] = {proton_mass_.value(), neutron_mass_.value()};
    real_type const dm[] = {target.proton_loss_energy().value(),
                            target.neutron_loss_energy().value()};
    static_assert(std::size(mass) == ZoneComponent::NucleonArray::size());
    static_assert(std::size(dm) == std::size(mass));
```
This is a little safer

**Review by sethrj — COMMENTED** [review]
One or two small suggestions; I'll let @amandalund review and then I can follow on after the merge with the numerical integrator.

**Review by amandalund — APPROVED** [review]
This looks great @whokion!

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [inline src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:388]: Does 
```suggestion
        if (soft_eq(result, integral))
```
not give the expected behavior?

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:171]: Not sure about this.  The number of zones is used for other properties (i.e., volumes, etc), so it may be still more readable keeping it as a common size across the whole code, which is called just once for a specific Isotope.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Okay, it is removed from the zone data (but keep the data structure as it was since more data may be added later).

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Okay. Changed as suggested.

File: src/celeritas/neutron/data/NeutronInelasticData.hh
Comment (whokion) [reply src/celeritas/neutron/data/NeutronInelasticData.hh]: Yes, this is now removed.

File: src/celeritas/neutron/data/NeutronInelasticData.hh
Comment (whokion) [reply src/celeritas/neutron/data/NeutronInelasticData.hh:98]: Actually, I originally did what you are suggesting, but I may want to reorganize data structure by grouping data separately needed for the process (i.e., macro_xs) and for the interactor (nucleon_xs, zone_data and other data which will be added later).  Although NuclearZoneData is too specific now and can be renamed by the more generic name as more data are added, I would keep the structure as it is now and  reorganize them later as more codes and data are added.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Removed.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh]: Removed.

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (whokion) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:388]: Okay, I also tried this earlier, which did not produce the same results compared to the current ones - as epsilon is the tolerance for the ratio, not on the absolute value.  Of course, we can re-adjust the value epsilon for the direct comparison with some optimization later.  It may be better to keep it as it is until we fully verify results with Geant4 - at this moment, these zone data are identical to those from Geant4 (which depends on the particular epsilon value), but certainly need more downstream codes for further validation and verification.

**Review by sethrj — APPROVED** [review]
Nice, thanks!

File: src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh
Comment (sethrj) [reply src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh:171]: Sorry, i didn't see this wasn't addressed. Are you expecting `NuclearZoneBuilder::num_nuclear_zones` to be called outside of that helper class? The number of zones is implicitly available through the components size, so the separate assumptions between `num_nuclear_zones` and `calc_zone_component` is a dangerous redundancy.

**Pull request merged**
